packages/app/src/cli/services/dev/extension/websocket/handlers.ts

```diff
@@ -0,0 +1,113 @@
+import {
+  EventType,
+  IncomingDispatchMessage,
+  OutgoingDispatchMessage,
+  OutgoingMessage,
+  SetupWebSocketConnectionOptions,
+} from './models.js'
+import {RawData, WebSocket, WebSocketServer} from 'ws'
+import {http, output} from '@shopify/cli-kit'
+import {Duplex} from 'stream'
+
+export function websocketUpgradeHandler(
+  wss: WebSocketServer,
+  options: SetupWebSocketConnectionOptions,
+): (req: http.IncomingMessage, socket: Duplex, head: Buffer) => void {
+  return (request, socket, head) => {
+    if (request.url !== '/extensions') {
+      return
+    }
+    output.debug(`Upgrading HTTP request to a websocket connection`)
+    wss.handleUpgrade(request, socket, head, getConnectionDoneHandler(wss, options))
+  }
+}
+
+export function getConnectionDoneHandler(wss: WebSocketServer, options: SetupWebSocketConnectionOptions) {
+  return (ws: WebSocket) => {
+    output.debug(`Websocket connection successfully established`)
+    const connectedPayload = {
+      event: 'connected',
+      data: options.payloadStore.getConnectedPayload(),
+      version: '3',
+    }
+    output.debug(output.content`Sending connected payload: ${output.token.json(connectedPayload)}`)
+    ws.send(JSON.stringify(connectedPayload))
+    ws.on('message', getOnMessageHandler(wss, options))
+  }
+}
+
+export function getOnMessageHandler(wss: WebSocketServer, options: SetupWebSocketConnectionOptions) {
+  return (data: RawData) => {
+    const jsonData = JSON.parse(data.toString())
+    const {event: eventType, data: eventData} = jsonData
+
+    output.debug(output.content`Received websocket message with event type ${eventType} and data:
+${output.token.json(eventData)}
+          `)
+
+    if (eventType === 'update') {
+      /**
+       * App updates must take precedence over extensions. Otherwise the websocket server
+       * will send an update to the client with missing app data and will cause the loading
+       * of extensions to fail.
+       */
+      const payloadStoreApiKey = options.payloadStore.getRawPayload().app.apiKey
+      const eventAppApiKey = eventData.app?.apiKey
+      if (eventData.app && payloadStoreApiKey === eventAppApiKey) {
```

Comment:
@pepicrft @alfonso-noriega @vividviolet If these don't match should we return early?

Right now if they don't match we might still update the extensions.  My guess is if the API Key doesn't match it's a signal we shouldn't trust anything about this message and we should bail early.

Reply:
It should return early and not continue processing the message

Reply:
David and I updated this.